persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java

```diff
@@ -94,37 +114,179 @@ public void beforeChange(
 
   @Override
   public void onChange(InetAddress endpoint, ApplicationState state, VersionedValue value) {
+    if (state == ApplicationState.STATUS) {
+      return;
+    }
+
     EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
     if (epState == null || Gossiper.instance.isDeadState(epState)) {
       return;
     }
 
-    addStargateNode(endpoint, epState);
+    if (!isStargateNode(epState)) {
+      return;
+    }
+
+    if (!joinCluster(endpoint, epState)) {
+      applyState(endpoint, state, value, epState);
+    }
   }
 
   @Override
-  public void onAlive(InetAddress endpoint, EndpointState state) {}
+  public void onAlive(InetAddress endpoint, EndpointState state) {
+    if (!isStargateNode(state)) {
+      return;
+    }
+    notifyUp(endpoint);
+  }
 
   @Override
-  public void onDead(InetAddress endpoint, EndpointState state) {}
+  public void onDead(InetAddress endpoint, EndpointState state) {
+    if (!isStargateNode(state)) {
+      return;
+    }
+    notifyDown(endpoint);
+  }
 
   @Override
   public void onRemove(InetAddress endpoint) {
+    leaveCluster(endpoint);
+  }
+
+  @Override
+  public void onRestart(InetAddress endpoint, EndpointState state) {}
+
+  private boolean joinCluster(InetAddress endpoint, EndpointState state) {
+    if (!liveStargateNodes.add(endpoint)) {
```

Comment:
when this situation can occur?

Reply:
This can happen in `onChange()` when the nodes already been added. In that case it should update the single endpoint's application state.